cmd/gateway-gcs.go

```diff
@@ -292,8 +292,48 @@ func (l *gcsGateway) ListBuckets() ([]BucketInfo, error) {
 	return b, nil
 }
 
-// DeleteBucket delete a bucket on GCS
+// DeleteBucket delete a bucket on GCS.
 func (l *gcsGateway) DeleteBucket(bucket string) error {
+	itObject := l.client.Bucket(bucket).Objects(l.ctx, &storage.Query{Delimiter: slashSeparator, Versions: false})
+	// We list the bucket and if we find any objects we return BucketNotEmpty error. If we
+	// find only "minio.sys.temp/" then we remove it before deleting the bucket.
+	gcsMinioPathFound := false
+	nonGCSMinioPathFound := false
+	for {
+		objAttrs, err := itObject.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return gcsToObjectError(traceError(err))
+		}
+		if isGCSPrefix(objAttrs.Prefix) {
+			gcsMinioPathFound = true
+			continue
+		}
+		nonGCSMinioPathFound = true
+		break
+	}
+	if nonGCSMinioPathFound {
+		return gcsToObjectError(traceError(BucketNotEmpty{}))
+	}
+	if gcsMinioPathFound {
```

Comment:
It is possible that an object may be uploaded to this bucket after the above check and will be accidentally deleted. This is incorrect behaviour, we shouldn't delete user uploaded objects and should return `BucketNotEmpty` error. 

To avoid this we can list objects that match prefix `.minio.sys.temp` in the loop below for deleting internal objects. `Bucket.Delete` will fail if an object was uploaded while we were removing internal objects, which is the right behaviour.